compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -291,6 +294,16 @@ trait SpaceLogic {
 class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   import tpd._
 
+  override def intersectAtomTypes(tp1: Type, tp2: Type) = {
+    def isOpen(tp: Type) =
+      !tp.classSymbol.is(Sealed | Final) && !tp.termSymbol.is(Module)
+
+    if ((tp1.classSymbol.is(Trait) || tp2.classSymbol.is(Trait))
+        && isOpen(tp1) && isOpen(tp2))
+      Typ(AndType(tp1, tp2), true)
```

Comment:
Normally, we expect programmers will not abuse AND-types in pattern matches. However, to be safer, it's better to handle the case if `tp1` and `tp2` are themselves AND-types if it doesn't significantly complicate the implementation.